apollo-router/src/plugins/telemetry/metrics/apollo/duration_histogram.rs

```diff
@@ -1,3 +1,4 @@
+use std::ops::AddAssign;
```

Comment:
Reading this particular commit, I'm honestly confused about what SingleFieldStat represents. You talk about how you can aggregate data from multiple subgraph requests... but you also just can have a single field referenced many times inside a single request? That said it also feels like SingleTypeStat/SingleFieldStat are never actually created outside of test code? So are they actually used for anything?

Reply:
Partly I think this comes down to the fact that in the original drop of this code we should have made `DurationHistogram` `add_assign`. The `Single*` structs were all there to allow transmitting of the raw data to the other side of a queue where the histogram lived. However as @SimonSapin has made `DurationHistogram` there is actually no need for any of these `Single*` structs, and we can remove them. Possibly in a followup ticket.

Reply:
I also find this confusing. I feel the whole set of `SingleFoo` structs is not bringing much and could be removed (using `Foo` directly instead).

As far as I can tell the `SingleContextualizedStats::per_type_stat` field as well as the `SingleTypeStat` and `SingleFieldStat` structs were indeed not used before this PR. Now they are created inside the `per_type_stat` function:

```rust
            per_type: &mut HashMap<String, SingleTypeStat>,
```
```rust
            let field_stat = per_type
                .entry(node.parent_type.clone())
                .or_default()
                .per_field_stat
                .entry(field_name.clone())
                .or_insert_with(|| metrics::apollo::studio::SingleFieldStat {
                    return_type: node.r#type.clone(), // not `Default::default()`’s empty string
                    errors_count: 0,
                    latency: Default::default(),
                    estimated_execution_count: 0.0,
                    requests_with_errors_count: 0,
                });
```

`HashMap::entry` returns an enum that represents either a vacant entry or an occupied entry within the map. If it’s vacant, `.or_default()` creates a empty `SingleTypeStat` through its impl of the `Default` trait. Either way, `or_default` returns a `&mut SingleTypeStat`.

Then on the `per_field_stat: HashMap<String, SingleFieldStat>` struct field we call `entry` again, but this time with `or_insert_with` which uses a callback instead of `Default` to create a new `SingleFieldStat` to fill vacant entries.

Reply:
Just to be clear my main concern here (other than cleanliness) is that the implication that the only reason you'd need multiple latencies for a field is "multiple subgraph fetches" and not "a field can be referenced many times within a single subgraph fetch" made me worry that there might be deeper problems somewhere. Worth thinking about!

Reply:
You’re right, multiple references in a single fetch is possible too. Multiple fetches was only the first thing that came to mind because I’d just written that `for` loop.